packages/editor-ui/src/components/RunData.vue

```diff
@@ -1199,47 +1167,59 @@ export default defineComponent({
 			const itemsLabel = itemsCount > 0 ? ` (${itemsCount} ${items})` : '';
 			return option + this.$locale.baseText('ndv.output.of') + (this.maxRunIndex + 1) + itemsLabel;
 		},
-		getDataCount(
+		getData(
 			runIndex: number,
 			outputIndex: number,
 			connectionType: ConnectionTypes = NodeConnectionType.Main,
-		) {
-			if (this.pinData) {
-				return this.pinData.length;
-			}
+		): INodeExecutionData[] {
+			let inputData: INodeExecutionData[] = [];
 
-			if (this.node === null) {
-				return 0;
+			if (this.node) {
+				inputData = this.getNodeInputData(
+					this.node,
+					runIndex,
+					outputIndex,
+					this.paneType,
+					connectionType,
+				);
 			}
 
-			const runData: IRunData | null = this.workflowRunData;
-
-			if (runData === null || !runData.hasOwnProperty(this.node.name)) {
-				return 0;
+			if (inputData.length === 0 || !Array.isArray(inputData)) {
+				return [];
 			}
 
-			if (runData[this.node.name].length <= runIndex) {
+			return inputData;
+		},
+		getPinDataOrInputData(inputData: INodeExecutionData[]): INodeExecutionData[] {
+			if (this.pinData && !this.isProductionExecutionPreview) {
+				return Array.isArray(this.pinData)
+					? this.pinData.map((value) => ({
+							json: value,
+					  }))
+					: [
+							{
+								json: this.pinData,
+							},
+					  ];
+			}
+			return inputData;
+		},
+		getDataCount(
+			runIndex: number,
+			outputIndex: number,
+			connectionType: ConnectionTypes = NodeConnectionType.Main,
+		) {
+			if (!this.node) {
 				return 0;
 			}
 
-			if (runData[this.node.name][runIndex].hasOwnProperty('error')) {
+			const runData: IRunData | null = this.workflowRunData;
+			if (runData?.[this.node.name][runIndex].hasOwnProperty('error')) {
 				return 1;
 			}
 
-			if (
-				!runData[this.node.name][runIndex].hasOwnProperty('data') ||
-				runData[this.node.name][runIndex].data === undefined
-			) {
-				return 0;
-			}
-
-			const inputData = this.getInputData(
-				runData[this.node.name][runIndex].data!,
-				outputIndex,
-				connectionType,
-			);
-
-			return inputData.length;
+			const inputData = this.getData(runIndex, outputIndex, connectionType);
+			return this.getPinDataOrInputData(inputData).length;
```

Comment:
`getPinDataOrInputData` potentially calls `map` on `pinData` but here we only care about the length, seems unnecessary work?

Reply:
You're right, I just went ahead a bit for the search functionality. Before returning the length I will have to do a filter on either the live data or pinned data to keep pagination in sync with search results.

I really hate I have to do it twice but I couldn't came a better solution at the moment.
My previous solution was better but it failed for paired items and if node branches where we need to know the length also in advance (because they are shown on the tabs - like 2 item(s))

So this is why `getData`and `getDataCount` is parametrised